### PR TITLE
Introduce `FluxCollectToImmutableList` Refaster rule

### PR DESCRIPTION
**Pull request by mlrprananta**
Based on internal inquiry

Suggested commit message:
```
Introduce `FluxCollectToImmutableList` Refaster rule (#570)
```

### CODE REVIEW

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/ReactorRules.java [hunk 1]
```diff
@@ -1166,6 +1167,23 @@ Flux<T> after(Flux<T> flux, Predicate<? super T> predicate, Comparator<? super T
     }
   }
 
+  /**
+   * Prefer {@link Flux#collect(Collector)} with {@link ImmutableList#toImmutableList} over {@link
+   * Flux#collectList()}.
+   */
+  static final class FluxCollectListToCollectImmutableList<T> {
```
Comment (mlrprananta) [inline error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/ReactorRules.java:1174]: Name might be a bit verbose, open to suggestions

Comment (mlrprananta) [reply error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/ReactorRules.java:1174]: Maybe just `FluxCollectToImmutableList`?

Comment (rickie) [reply error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/ReactorRules.java:1174]: Yes that is the perfect suggestions. We base the name on the content of the after template :). Will push the suggestion.

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/ReactorRules.java [hunk 2]
```diff
@@ -1166,6 +1167,23 @@ Flux<T> after(Flux<T> flux, Predicate<? super T> predicate, Comparator<? super T
...
+    @BeforeTemplate
+    Mono<List<T>> before(Flux<T> flux) {
```
Comment (rickie) [inline error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/ReactorRules.java:1176]: We can add: `flux.collect(toImmutableList()).map(ImmutableList::copyOf);`.

**Review by rickie — APPROVED** [review]
Added a commit with an extra rule :). Changes LGTM!

Slightly tweaked the suggested commit message.

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/ReactorRules.java [hunk 3]
```diff
@@ -1168,15 +1168,20 @@ Flux<T> after(Flux<T> flux, Predicate<? super T> predicate, Comparator<? super T
...
-  static final class FluxCollectListToCollectImmutableList<T> {
+  static final class FluxCollectToImmutableList<T> {
     @BeforeTemplate
     Mono<List<T>> before(Flux<T> flux) {
       return flux.collectList();
     }
 
+    @BeforeTemplate
+    Mono<List<T>> before2(Flux<T> flux) {
+      return flux.collect(toImmutableList()).map(ImmutableList::copyOf);
```
Comment (mlrprananta) [inline error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/ReactorRules.java:1182]: Can we use `Refaster.anyOf` in `before` instead?

Comment (rickie) [reply error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/ReactorRules.java:1182]: Ah yeah, same return type so we definitely can 👍🏻.

Comment (mlrprananta) [reply error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/ReactorRules.java:1182]: Will push

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/ReactorRules.java [hunk 4]
```diff
@@ -1166,6 +1167,24 @@ Flux<T> after(Flux<T> flux, Predicate<? super T> predicate, Comparator<? super T
...
+    @BeforeTemplate
+    Mono<List<T>> before(Flux<T> flux) {
+      return Refaster.anyOf(
+          flux.collectList(), flux.collect(toImmutableList()).map(ImmutableList::copyOf));
+    }
+
+    @AfterTemplate
+    @UseImportPolicy(STATIC_IMPORT_ALWAYS)
+    Mono<List<T>> after(Flux<T> flux) {
+      return flux.collect(toImmutableList());
```
Comment (Stephan202) [inline error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/ReactorRules.java]: Let's make explicit that _strictly speaking_ this expression has a slightly different return type.

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/ReactorRules.java (hunk 2 above)
Comment (Stephan202) [reply error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/ReactorRules.java:1183]: Hmm, the `((Mono<ImmutableList<T>>) flux).map(ImmutableList::copyOf)` simplification should be in a rule of its own. (And is in fact a special case of a large range of rules.)

**Review by Stephan202 — APPROVED** [review]
Rebased and added a commit.

Suggested commit message:
```
Introduce `FluxCollectToImmutableList` Refaster rule (#570)

And extend `MonoIdentity` to simplify `mono.map(ImmutableList::copyOf)`
expressions where possible, as the new rule may introduce such cases.
```

**Comment by rickie** [comment]
Changes LGTM! Will merge once 🟢.

**Pull request merged**